fixes #3697, #3701 - more comprehensive field validations

stbenjam opened the pull request:
- Hostnames should be validated per RFC1123 
- Hostname shortnames should not have a period
- Upcase hostnames automatically downcased instead of outright failure
- IPv4 addresses validated that each octet is < 255

stbenjam committed:
fixes #3697, #3701 - more comprehensive field validations

domcleal commented on app/models/nic/base.rb:
```diff
@@ -10,9 +10,15 @@ class Base < ActiveRecord::Base
+                     :presence => true,
+                     :format => {:with => Net::Validations::MAC_REGEXP}
+
+    validates :name, :allow_blank => true,
```
This should be in managed.rb as we support "unmanaged" NICs (with no name) and managed with a DNS entry etc.

domcleal commented on app/models/nic/base.rb:
```diff
@@ -54,5 +60,9 @@ def uniq_with_hosts
     end
+
+    def normalize_name
+      self.name = Net::Validations.normalize_hostname(name) if self.name
```
maybe use `self.name.present?` which won't then run this on empty strings either (ditto in the host above)

domcleal commented on lib/net/validations.rb:
```diff
@@ -1,8 +1,10 @@
-    IP_REGEXP  = /^(\d{1,3}\.){3}\d{1,3}$/
+    IP_REGEXP  = /^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(\.|$)){4}$/
     MAC_REGEXP = /^([a-f0-9]{1,2}:){5}[a-f0-9]{1,2}$/i
+    HOST_REGEXP = /^(([a-z0-9]|[a-z0-9][a-z0-9\-]*[a-z0-9])\.)*([a-z0-9]|[a-z0-9][a-z0-9\-]*[a-z0-9])$/
```
Nitpick, these should use \A instead of ^ and \Z instead of $ (difference is end of string instead of end of line)

stbenjam replied:
Ah, ok, didn't realize the difference. Should be lowercase 'z',  though? 

http://stackoverflow.com/questions/577653/difference-between-a-z-and-in-ruby-regular-expressions

domcleal replied:
You're right, it should... that's an error in a6b0eeb0 too then :(

GregSutcliffe commented on lib/net/validations.rb:
```diff
@@ -18,6 +20,12 @@ def validate_mac mac
+    # validates the hostname
+    def validate_hostname hostname
+      raise Error, "Invalid Hostname #{hostname}" unless (hostname =~ HOST_REGEXP)
+      mac
```
Why are we returning a mac here? typo?

stbenjam replied:
Yup, copy-paste from the other method.  This is only used in ./test/lib/net/validations_test.rb, which I actually need to update to include a test for hostname validation.

GregSutcliffe commented on test/unit/nic_test.rb:
```diff
@@ -19,7 +19,7 @@ def teardown
   test "type casting should return the correct class" do
     i = ''
-    i = Nic::Base.create! :ip => "127.2.3.8", :mac => "babbccddeeff", :host => hosts(:one), :name => hosts(:one).name + "!", :type => "Nic::Interface"
+    i = Nic::Base.create! :ip => "127.2.3.8", :mac => "babbccddeeff", :host => hosts(:one), :name => hosts(:one).name, :type => "Nic::Interface"
```
why drop the "!"?

stbenjam replied:
Why was it there to begin with?  

The original code added a NIC with a hostname "my5name.mydomain.net!" -- which of course doesn't validate now that we're checking.

GregSutcliffe replied:
Fair enough, I'm always wary of test changes that seem unrelated at first glance :)

GregSutcliffe commented on test/unit/subnet_test.rb:
```diff
@@ -225,6 +225,9 @@ def setup_user operation
     refute s.valid?
+    # greater than 255
+    s.network = "300.300.300.0"
+    refute s.valid?
```
Good idea, but I don't see any changes to the subnet model to use the new validations?

GregSutcliffe replied:
Scratch that, I see it's in the model already :P

GregSutcliffe commented:
Looks ok to me, once that typo is fixed up :)

stbenjam commented on app/models/host/managed.rb:
```diff
@@ -154,6 +155,11 @@ def shortname
     domain.nil? ? name : name.chomp("." + domain.name)
```
This breaks: 
  https://github.com/theforeman/foreman/blob/develop/test/lib/has_many_common_test.rb#L59
1. Given a host, my5name.mydomain.net
2. Set domain directly to "yourdomain.net" (as is done in has_many_common_test)
3. shortname_has_no_periods gets called on host.save! 
4. But "name" is STILL "my5name.mydomain.net" but domain is "yourdomain.net"
5. So the "shortname" is my5name.mydomain.net"
6. Validation fails -- there's periods in the "shortname"

Any ideas here how to get name updated to reflect the new real domain?

domcleal replied:
Sorry for the delay!

I think the way to solve this might be to add a before_validation method which is able to update the name when the domain changes, rather than overriding the setters (there are more than just domain_name=).

Rails provides old_\* methods like old_domain, so you can compare old_domain to domain in the before_validation, then strip off the old domain.

stbenjam replied:
Great, that sounds better.  I suspected there was a more rails idiomatic way to do this.

stbenjam commented on app/models/host/managed.rb:
```diff
@@ -154,6 +155,19 @@ def shortname
+    # the old one by setting :name to shortname
+    update_attribute(:name, shortname)
+    update_attribute(:domain, Domain.find_by_name(new_domain))
+  end
```
So this fixes the problem I mentioned earlier with the test (changes the domain by string).  

I don't know if this is too underhanded or bad practice to override domain_name here.  This is approaching the limit of my knowledge.

domcleal commented on app/models/host/managed.rb:
```diff
@@ -154,6 +155,19 @@ def shortname
 
+  def shortname_has_no_periods
+    return if name.nil?
+    errors.add(:name, "Do not include periods except in the domain name") if shortname.include? "."
```
Although this looks ok in the UI, it will be odd in the API etc where the full validation messages are shown as it'll do:

"name Do not include..."

Best to treat it as a sentence beginning with the field name "name", e.g. `errors.add(:name, "must not include periods..")`

Also surround the message with `_( )` so it's translated.

domcleal commented on app/models/host/managed.rb:
```diff
@@ -154,6 +155,19 @@ def shortname
     domain.nil? ? name : name.chomp("." + domain.name)
   end
 
+  def shortname_has_no_periods
```
I'm a bit worried about unmanaged hosts that we pick up through reports, since they might not have a domain set, so shortname might not strip off the domain - causing this validation to fail.  @GregSutcliffe?

stbenjam replied:
Would unmanaged hosts hit this in managed.rb?

domcleal replied:
I think they're still Host::Managed instances, but with managed=false :)

GregSutcliffe replied:
I'm not sure either. Can we add a test for that? then we'll know for sure :)

stbenjam commented:
I think I've got it all sorted.  We'll see if Jenkins explodes.

https://github.com/theforeman/foreman/pull/1041/files#diff-2d53894341fd1cf84a823e3bd0855495L752 was  broken -- the name wasn't actually being updated.  I've fixed it, and added a test.  That fixed my issue.

I also only do the shortname validation on managed hosts.  Unmanaged hosts with domains we don't know about could conceivably report, right?

domcleal commented:
@stbenjam indeed they could

stbenjam commented:
@domcleal Ok -- so I added tests to cover that.  

It also has the side benefit of testing successful importing from facts and reports, which I don't see tested anywhere else (only the negative case was tested).

domcleal commented on app/models/host/managed.rb:
```diff
@@ -746,13 +745,17 @@ def normalize_hostname
       if !new_record? and changed_attributes['domain_id'].present?
         old_domain = Domain.find(changed_attributes["domain_id"])
-        self.name.gsub(old_domain.to_s,"")
+        update_attribute(:name, self.name.chomp("." + old_domain.to_s))
```
Just setting `self.name =` here might better, update_attribute causes an immediate DB update.

domcleal commented on app/models/host/managed.rb:
```diff
@@ -746,13 +745,17 @@ def normalize_hostname
     end
+
+    # A managed host we should know the domain for; and the shortname shouldn't include a period
+    errors.add(:name, "must not include periods") if managed? and shortname.include? "."
```
use `_("must not..")` for i18n

domcleal commented on test/unit/host_test.rb:
```diff
@@ -131,6 +153,19 @@ class HostTest < ActiveSupport::TestCase
+    Setting[:create_new_host_when_facts_are_uploaded] =
+        Setting.find_by_name("create_new_host_when_facts_are_uploaded").default
+    assert_equal host.name, 'sinn1636.lan'
+    assert_equal old_count + 1, Host.count
```
```
assert_difference "Host.count", 1 do
  ...
end
```

will do the same (1's the default too, you can leave it off)

domcleal commented on test/unit/usergroup_test.rb:
```diff
@@ -18,19 +18,19 @@ class UsergroupTest < ActiveSupport::TestCase
 
   test "name is unique across user as well as usergroup" do
-    user = User.create :auth_source => auth_sources(:one), :login => "user", :mail  => "user@someware.com"
+    user = User.create :auth_source => auth_sources(:one), :login => "user", :mail  => "user@mydomain.net"
```
Is this change necessary?

stbenjam replied:
Not anymore.

domcleal commented on lib/net/validations.rb:
```diff
@@ -45,5 +53,9 @@ def self.normalize_mac mac
     end
+
+   def self.normalize_hostname hostname
+     hostname.downcase if hostname.present?
```
It's a minor thing, but this method will return nil if hostname is "".  It might be better to write:

```
hostname.downcase! if hostname.present?
hostname
```

domcleal commented:
Thanks @stbenjam!  Merged as 4bae5ced for Foreman 1.4.0.

Just to note, with @witlessbird's help, we changed the tests to call `@host.valid?` and check the error messages so we're specifically checking for which validations failed (and the error message) instead of just checking the host wouldn't save, just in case another validation failure would prevent it.

stbenjam commented:
Oh right, :+1: thanks. I didn't think testing host.save all the way through.
